dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,263 @@
+        case RPC_S_SERVER_UNAVAILABLE:
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    inline bool IsPackagedProcess()
+    {
+        wchar_t packageFullName[PACKAGE_FULL_NAME_MAX_LENGTH + 1] = {};
+
+        UINT32 packageFullNameLength = static_cast<UINT32>(ARRAYSIZE(packageFullName));
+
+        const auto packagedProcessHResult = ::GetCurrentPackageFullName(&packageFullNameLength, packageFullName);
+
+        if (packagedProcessHResult == APPMODEL_ERROR_NO_PACKAGE)
+        {
+            return false;
+        }
+
+        THROW_IF_WIN32_ERROR(packagedProcessHResult);
+
+        return true;
+    }
+
+    winrt::IAsyncOperationWithProgress<winrt::Microsoft::Windows::PushNotifications::PushNotificationCreateChannelResult, winrt::Microsoft::Windows::PushNotifications::PushNotificationCreateChannelStatus> PushNotificationManager::CreateChannelAsync(winrt::guid remoteId)
+    {
+        static bool s_remoteIdInProgress;
+        static wil::critical_section s_lock;
+
+        THROW_HR_IF(E_INVALIDARG, (remoteId == winrt::guid()));
+
+        // API supports channel requests only for packaged applications for v0.8 version
+        THROW_HR_IF(E_NOTIMPL, (IsPackagedProcess() == false));
+
+        winrt::Microsoft::Windows::PushNotifications::PushNotificationCreateChannelResult channelResult{ nullptr };
+        {
+            auto lock = s_lock.lock();
+            if (s_remoteIdInProgress == false)
+            {
+                s_remoteIdInProgress = true;
+            }
+            else
+            {
+                channelResult = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationCreateChannelResult>(
+                    nullptr, WPN_E_OUTSTANDING_CHANNEL_REQUEST, PushNotificationChannelStatus::CompletedFailure);
+                co_return channelResult;
+            }
+        }
+
+        auto scopeExit = wil::scope_exit([&]()
+            {
+                auto lock = s_lock.lock();
+                s_remoteIdInProgress = false;
+            });
+
+        auto cancellation{ co_await winrt::get_cancellation_token() };
+
+        cancellation.enable_propagation(true);
+
+        // Allow to register the progress and complete handler
+        co_await resume_background();
+
+        auto progress{ co_await winrt::get_progress_token() };
+
+        uint8_t retryCount = 0;
+        winrt::hresult channelRequestResult = E_PENDING;
+        PushNotificationChannelStatus status = PushNotificationChannelStatus::InProgress;
+
+        PushNotificationCreateChannelStatus
+            channelStatus = { channelRequestResult, status, retryCount };
+
+        progress(channelStatus);
+
+        PushNotificationChannelManager channelManager{};
+        winrt::PushNotificationChannel pushChannelReceived{ nullptr };
+
+        for (auto backOffTimeInSeconds = c_minBackoffSeconds; backOffTimeInSeconds <= c_maxBackoffSeconds * 2; backOffTimeInSeconds *= 2)
```

Comment:
Everytime we enter the loop we request the channel, if fail, wait for the backofftimeinseconds and retry. If for whatever reason we waited for 16mins which is maxoff then after waiting we want to request the channel for one time more so we enter the loop using c_maxBackoffSeconds * 2. Thanks.

Reply:
Okay, I see now. We want to make one last try after reaching the maximum backoff, so we intentionally allow a "too large" backoff through, but then detect it before we do the `resume_after`.  This is fragile. The backoff variable should hold the backoff.

Also, should probably use `std::chrono` time durations rather than encoding seconds in integers. More readable, and avoids conversions.

```cpp
    inline constexpr auto c_initialBackoff = 30s;
    inline constexpr auto c_maxBackoff = 16min;
    inline constexpr auto backoffFactor = 2;

    auto backoff = c_initialBackoff;

    for (;;)
    {
        try
        {
            ....
            co_return success;
        }
        catch (...)
        {
            if (unrecoverable || backoff > c_maxBackoff)
            {
                co_return failed;
            }
        }
        co_await resume_after(backoff);
        backoff = backoff * backoffFactor;
    }
...

Reply:
Also, 16 minutes seems absurd. Nobody is going to wait 16 minutes for an operation to complete.

Reply:
@oldnewthing would also like to take your suggestion on the appropriate max duration that would be ideal where we can continue to wait and request channel before giving up?

Reply:
Why retry at all? Just try once. If the caller wants to retry, they can choose how often to retry and for how long.

Reply:
@oldnewthing  So when we compiled the list of errors from channel requests, we saw that the majority were retryable errors. So we wanted to retry these requests on behalf of the caller and improve the reliability of channel requests.

Reply:
Yes. The rationale was that the complexities of retryable errors and non-retryable errors should be abstracted away from the developer and the reunion layer handles the complexity. Sometimes, network goes down or the socket with the end point could be busted and eventually recovers. We don't want to have the developer figure that out.

Having said that, I think we need to max it out to something like 5 retries with an interval of 1 minute. Give up after 5 minutes if we keep failing.

Reply:
thanks pavan. addressed it.